design-proposals/vhostuser-interface.md

```diff
@@ -0,0 +1,78 @@
+
+- As a result, when we have a host with DPDK dataplane running in Userspace, the traffic from kubevirt VM will go through the kernel and reach the dataplane running in userspace. This longer path slows down the traffic between the application and the dataplane when both are running in userspace.
+
+- In order to improve the traffic performance and make use of the dpdk, we need to use a different path that can by-pass the kernel and reach the dataplane from the user space directly. This so called fast path will run within the user space and will allow fast packet processing.
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a valid CNI is being used.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. 
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. A client posts a new VM definition to the K8s API Server.  
+    2. The K8s API Server validates the input and creates a VM custom resource definition (CRD) object.  
+    3. The virt-controller observes the creation of the new VM object and creates a corresponding virt-launcher pod.
+    4. **The virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
+    5. Scheduling of the launcher pod will be handled by kubernetes.
+    6. **The CNI will mount the shared-dir i.e., EmptyDir volume with the mount-path as "/var/lib/cni/usrcni/".**
+    7. **The CNI will update the virt-launcher pod annotations with vhostsocket-file name and details.**
+    8. The virt-controller observes the creation of the virt-launcher pod for the VM and updates the nodeName field in the VM object with the nodeName retrieved from the virt-launcher pod.
```

Comment:
I don't think it is necessary to list all the steps that don't diverge from the regular flow. For example , the 8 step isn't very clear to me